ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/query/StateBackupService.java

```diff
@@ -0,0 +1,469 @@
+    }
+  }
+
+  private TrieIterator.State visitAccount(final Bytes32 nodeKey, final Node<Bytes> node) {
+    if (node.getValue().isEmpty()) {
+      return State.CONTINUE;
+    }
+
+    backupStatus.currentAccount = nodeKey;
+    final Bytes nodeValue = node.getValue().orElse(Hash.EMPTY);
+    final StateTrieAccountValue account =
+        StateTrieAccountValue.readFrom(new BytesValueRLPInput(nodeValue, false));
+
+    final Bytes code = worldStateStorage.getCode(account.getCodeHash()).orElse(Bytes.EMPTY);
+    backupStatus.codeSize.addAndGet(code.size());
+
+    final BytesValueRLPOutput accountOutput = new BytesValueRLPOutput();
+    accountOutput.startList();
+    accountOutput.writeBytes(nodeKey); // trie hash
+    accountOutput.writeBytes(nodeValue); // account rlp
+    accountOutput.writeBytes(code); // code
+    accountOutput.endList();
+
+    try {
+      accountFileWriter.writeBytes(accountOutput.encoded().toArrayUnsafe());
+    } catch (final IOException ioe) {
+      LOG.error("Failure writing backup", ioe);
+      return State.STOP;
+    }
+
+    // storage is written for each leaf, otherwise the whole trie would have to fit in memory
+    final StoredMerklePatriciaTrie<Bytes32, Bytes> storageTrie =
+        new StoredMerklePatriciaTrie<>(
+            worldStateStorage::getAccountStateTrieNode,
+            account.getStorageRoot(),
+            Function.identity(),
+            Function.identity());
+    storageTrie.visitLeafs(
+        (storageKey, storageValue) ->
+            visitAccountStorage(storageKey, storageValue, accountFileWriter));
+
+    try {
+      accountFileWriter.writeBytes(ACCOUNT_END_MARKER.toArrayUnsafe());
+    } catch (final IOException ioe) {
+      LOG.error("Failure writing backup", ioe);
+      return State.STOP;
+    }
+
+    backupStatus.accountCount.incrementAndGet();
+    return State.CONTINUE;
+  }
+
+  private void backupChaindata() throws IOException {
+    try (final RollingFileWriter headerWriter =
+            new RollingFileWriter(this::headerFileName, backupStatus.compressed);
+        final RollingFileWriter bodyWriter =
+            new RollingFileWriter(this::bodyFileName, backupStatus.compressed);
+        final RollingFileWriter receiptsWriter =
+            new RollingFileWriter(this::receiptFileName, backupStatus.compressed)) {
+      for (int blockNumber = 0; blockNumber <= backupStatus.targetBlock; blockNumber++) {
+        final Optional<Block> block = blockchain.getBlockByNumber(blockNumber);
+        if (block.isEmpty()) {
+          throw new IllegalStateException(
+              "Block data for " + blockNumber + " was not found in the archive");
+        }
+        final Optional<List<TransactionReceipt>> receipts =
+            blockchain.getTxReceipts(block.get().getHash());
+        if (receipts.isEmpty()) {
+          throw new IllegalStateException(
+              "Receipts for " + blockNumber + " was not found in the archive");
+        }
```

Comment:
It will return an optional of an empty list if the block header exists and there are no transacitons.  If the block doesn't exist we get an empty optional.